File: frontend/src/app/core/services/curso-asignado.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { CursoAsignado } from './docente.service';

export interface CursoAsignadoRequestDTO {
  docenteId: number;
  aulaId: number;
  areaCurricular: string;
  anioAcademico: number;
  diaSemana?: string;
  horaInicio?: string;
  horaFin?: string;
}

export interface HorarioRequestDTO {
  diaSemana: string;
  horaInicio: string;
  horaFin: string;
}

@Injectable({
  providedIn: 'root'
})
export class CursoAsignadoService {
  private http = inject(HttpClient);
  private readonly BASE = `${environment.apiUrl}/cursos-asignados`;

  crear(dto: CursoAsignadoRequestDTO): Observable<CursoAsignado> {
    return this.http.post<CursoAsignado>(this.BASE, dto);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.BASE}/${id}`);
  }

  /** Lista las asignaciones (con horario, si ya fue definido) de una aula. */
  listarPorAula(aulaId: number): Observable<CursoAsignado[]> {
    return this.http.get<CursoAsignado[]>(`${this.BASE}/aula/${aulaId}`);
  }

  /** Fija o actualiza el horario (día + hora) de una asignación ya creada. */
  actualizarHorario(id: number, dto: HorarioRequestDTO): Observable<CursoAsignado> {
    return this.http.patch<CursoAsignado>(`${this.BASE}/${id}/horario`, dto);
  }
}

File: backend/src/main/java/com/colegio/bastidas/service/CursoAsignadoService.java
package com.colegio.bastidas.service;

import com.colegio.bastidas.dto.curso.CursoAsignadoRequestDTO;
import com.colegio.bastidas.dto.curso.HorarioRequestDTO;
import com.colegio.bastidas.model.CursoAsignado;

import java.util.List;

/**
 * Contrato de negocio para la asignación de cursos (Docente + Aula + Área) a cargo del Director.
 */
public interface CursoAsignadoService {

    /** Asigna un docente a un aula/área/año. Falla si ya existe la misma asignación o hay solape de horario. */
    CursoAsignado crear(CursoAsignadoRequestDTO dto);

    /** Elimina una asignación existente. */
    void eliminar(Long id);

    /** Lista las asignaciones (con horario, si ya se definió) de una aula. */
    List<CursoAsignado> listarPorAula(Long aulaId);

    /** Fija o actualiza el horario (día + hora) de una asignación ya creada. */
    CursoAsignado actualizarHorario(Long id, HorarioRequestDTO dto);
}
